lib/srv/db/postgres/users.go

```diff
@@ -81,8 +82,127 @@ func (e *Engine) ActivateUser(ctx context.Context, sessionCtx *common.Session) e
+		return trace.Wrap(err)
+	}
 	return nil
+}
+
+type TablePermission struct {
+	Privilege string `json:"privilege"`
+	Schema    string `json:"schema"`
+	Table     string `json:"table"`
+}
+
+type Permissions struct {
+	Tables []TablePermission `json:"tables"`
+}
+
+var pgTablePerms = map[string]struct{}{
+	"SELECT":     {},
+	"INSERT":     {},
+	"UPDATE":     {},
+	"DELETE":     {},
+	"TRUNCATE":   {},
+	"REFERENCES": {},
+	"TRIGGER":    {},
+}
+
+func checkPgPermission(objKind, perm string) error {
+	// for now, only tables are supported. ignore other kinds of objects.
+	if objKind != permissions.ObjectKindTable {
+		return nil
+	}
+
+	normalized := strings.ToUpper(strings.TrimSpace(perm))
+	_, found := pgTablePerms[normalized]
+	if !found {
+		return trace.BadParameter("unrecognized %q Postgres permission: %q", objKind, perm)
+	}
+	return nil
+}
+
+// convertPermissions converts the permissions into a stable format expected by the stored procedure. It also filters out any unsupported objects.
+func convertPermissions(perms permissions.PermissionSet) (*Permissions, error) {
+	var out Permissions
+	var errors []error
+	for permission, objects := range perms {
+		for _, obj := range objects {
+			if err := checkPgPermission(obj.GetSpec().ObjectKind, permission); err != nil {
+				errors = append(errors, err)
+				continue
+			}
+			if obj.GetSpec().ObjectKind == permissions.ObjectKindTable {
+				out.Tables = append(out.Tables, TablePermission{
+					Privilege: permission,
+					Schema:    obj.GetSpec().Schema,
+					Table:     obj.GetSpec().Name,
+				})
+			}
+		}
+	}
+	if len(errors) > 0 {
+		return nil, trace.NewAggregate(errors...)
+	}
+	return &out, nil
+}
+
+func (e *Engine) applyPermissions(ctx context.Context, sessionCtx *common.Session, conn *pgx.Conn) error {
+	allow, _ := sessionCtx.Checker.GetDatabasePermissions()
+	if len(allow) == 0 {
+		e.Log.Infof("Skipping applying fine-grained permissions: none to apply.")
+		return nil
+	}
+
+	if len(sessionCtx.DatabaseRoles) > 0 {
+		e.Log.Errorf("Cannot apply fine-grained permissions: non-empty list of database roles (%v).", sessionCtx.DatabaseRoles)
+		return trace.BadParameter("fine-grained database permissions and database roles are mutually exclusive, yet both were provided.")
+	}
+
+	rules, err := e.AuthClient.GetDatabaseObjectImportRules(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
 
+	objsFetched, err := fetchDatabaseObjects(ctx, sessionCtx, conn)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	counts, _ := permissions.CountObjectKinds(objsFetched)
+	e.Log.Infof("Fetched %v objects from the database (%v).", len(objsFetched), counts)
+
+	objsTagged := permissions.ApplyDatabaseObjectImportRules(rules, sessionCtx.Database, objsFetched)
+	counts, _ = permissions.CountObjectKinds(objsTagged)
+	e.Log.Infof("Tagged %v database objects (%v).", len(objsTagged), counts)
+
+	permissionSet, err := permissions.CalculatePermissions(sessionCtx.Checker, objsTagged)
```

Comment:
Does `CalculatePermissions` have the same problem we discussed offline? like would it grab allows/denies from all roles (vs the role that matches this database)?

Reply:
Done, except for the tests which I'm working on.

Reply:
Done.